examples/configmapgen_controller.rs

```diff
@@ -112,16 +116,45 @@ async fn main() -> Result<()> {
     let cmgs = Api::<ConfigMapGenerator>::all(client.clone());
     let cms = Api::<ConfigMap>::all(client.clone());
 
-    Controller::new(cmgs, ListParams::default())
-        .owns(cms, ListParams::default())
-        .run(reconcile, error_policy, Context::new(Data { client }))
-        .for_each(|res| async move {
-            match res {
-                Ok(o) => info!("reconciled {:?}", o),
-                Err(e) => warn!("reconcile failed: {}", Report::from(e)),
-            }
-        })
-        .await;
+    log::info!("starting configmapgen-controller");
+    log::info!("press <enter> to force a reconciliation of all objects");
+    log::info!("press <ctrl>+c to shut down gracefully");
+
+    let (mut reload_tx, reload_rx) = futures::channel::mpsc::channel(0);
+    // Using a regular background thread since tokio::io::stdin() doesn't allow aborting reads,
+    // and its worker prevents the Tokio runtime from shutting down.
+    std::thread::spawn(move || {
+        for _ in std::io::BufReader::new(std::io::stdin()).lines() {
+            let _ = reload_tx.try_send(());
+        }
+    });
+
+    let (graceful_shutdown_tx, graceful_shutdown_rx) = futures::channel::oneshot::channel();
+    let forceful_shutdown = async {
+        tokio::signal::ctrl_c().await.unwrap();
+        graceful_shutdown_tx.send(()).unwrap();
+        log::info!("starting graceful shutdown, press <ctrl>+c again to shut down forcefully");
+        tokio::signal::ctrl_c().await.unwrap();
+        log::info!("forceful shutdown requested");
+        std::process::exit(0);
+    };
 
+    futures::future::select(
+        Controller::new(cmgs, ListParams::default())
+            .owns(cms, ListParams::default())
+            .reconcile_all_on(reload_rx.map(|_| ()))
+            .graceful_shutdown_on(graceful_shutdown_rx.map(|_| ()))
```

Comment:
while I think it's great that we are able to expose this, it's also a bit boilerplatey for the standard use case. if we are targetting a kubernetes deployed controller then the shutdown signal is always going to be `SIGTERM`.

i think the the method here makes complete sense for configurability, but maybe we could also have a `Controller::install_sigterm_handler()` method that sets this up the same thing under the hood using `tokio::signal` inside the building step of the `Controller`?

Reply:
How about `shutdown_on_sigint` or `shutdown_on_ctrl_c`?

Reply:
i feel `ctrl_c` and `sigint` naming is strange if we are putting it inside kubernetes which sends `SIGTERM`.
but it looks like we need to use `SIGINT` locally?

Reply:
Yeah, in-cluster (or when running in systemd or similar) we want SIGTERM, when running from cargo we want SIGINT (or whatever the Windows equivalent is). We _should_ be safe if we just treat the two as equivalent.

Reply:
yeah, if we can get tokio::signal to listen on both, then that would be ideal.

Reply:
How about `Controller::manage_termination_signals()` and make it trigger on either `SIGINT` or `SIGTERM`?

Interestingly, [actix-web has an opt-out for this scenario](https://docs.rs/actix-web/3.3.2/actix_web/struct.HttpServer.html#method.disable_signals). That might also be somewhere to go to, but that's probably too early.

Reply:
IMO installing signal handlers without asking for permission is a bit presumptuous if we aren't sure that we own the whole process. If we had a `#[kube_runtime::main]` or similar I'd be for it, but not in the current state.

Reply:
Added a helper now for managing the signal handling for you.